### PR TITLE
tests: add application for release specs 6

### PR DESCRIPTION
**Pull request by miri64**
This PR adds a test to allow for testing of [release specs 6](https://github.com/RIOT-OS/Release-Specs/blob/master/06-single-hop-udp/06-single-hop-udp.md)

### CODE REVIEW

**Comment by kYc0o** [comment]
Can you just address Murdock errors? BTW, Murdock2 says OK and Murdock don't, this is strange no?

**Comment by miri64** [comment]
It says the RAM is overflowing in gnrc_networking (which I did not touch oO).

File: tests/gnrc_udp/Makefile [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# name of your application
+APPLICATION = gnrc_networking
```
Comment (miri64) [inline tests/gnrc_udp/Makefile]: Maybe that is the reason ^^"

**Comment by miri64** [comment]
It's still rather strange, that Murdock and Murdock2 come to different results here.

**Comment by miri64** [comment]
@kaspar030 any idea? is this a false positive in Murdock?

**Comment by kaspar030** [comment]
hm. do you remember which platform the test failed? murdock2 should have the build log. if that looks ok, the build was probably ok, too...

**Comment by kYc0o** [comment]
AFAIR at least on nucleo-f103.

**Comment by miri64** [comment]
Mhhh... when I search for `tests/gnrc_udp` in the output of Murdock2 I don't find it....

File: tests/gnrc_udp/Makefile [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# name of your application
+APPLICATION = gnrc_udp
+
+# If no BOARD is found in the environment, use this default:
+BOARD ?= native
+
+# This has to be the absolute path to the RIOT base directory:
+RIOTBASE ?= $(CURDIR)/../../RIOT
```
Comment (kaspar030) [inline tests/gnrc_udp/Makefile]: This path is wrong...

**Comment by kaspar030** [comment]
> Mhhh... when I search for tests/gnrc_udp in the output of Murdock2 I don't find it....

```RIOTBASE``` is defined wrongly, leading to broken output of ```make info-boards-supported```.
I don't really understand why Murdock1 can build this.

File: tests/gnrc_udp/Makefile (hunk 2 above)
Comment (kaspar030) [inline tests/gnrc_udp/Makefile]: Please include the common tests Makefile.

File: tests/gnrc_udp/main.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
```
Comment (kaspar030) [inline tests/gnrc_udp/main.c]: That copyright correct?

File: tests/gnrc_udp/main.c [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     examples
+ * @{
+ *
+ * @file
+ * @brief       Example application for demonstrating the RIOT network stack
```
Comment (kaspar030) [inline tests/gnrc_udp/main.c]: Copy pasta

File: tests/gnrc_udp/main.c [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
```
Comment (kaspar030) [inline tests/gnrc_udp/main.c]: this correct?

File: tests/gnrc_udp/main.c [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    { "udp", "send data over UDP and listen on UDP ports", udp_cmd },
+    { NULL, NULL, NULL }
+};
+
+int main(void)
+{
+    /* we need a message queue for the thread running the shell in order to
+     * receive potentially fast incoming networking packets */
+    msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
+    puts("RIOT network stack example application");
```
Comment (kaspar030) [inline tests/gnrc_udp/main.c]: make more descriptive or remove

File: tests/gnrc_udp/udp.c [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    reply.content.value = (uint32_t)(-ENOTSUP);
+    reply.type = GNRC_NETAPI_MSG_TYPE_ACK;
+
+    while (1) {
+        msg_receive(&msg);
+
+        switch (msg.type) {
+            case GNRC_NETAPI_MSG_TYPE_RCV:
+                printf("Packets received: %d\n", ++rcv_count);
+                gnrc_pktbuf_release((gnrc_pktsnip_t *)msg.content.ptr);
```
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: cast not necessary

File: tests/gnrc_udp/udp.c [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    ipv6_addr_t addr;
+    size_t data_len;
+
+    /* parse destination address */
+    if (ipv6_addr_from_str(&addr, addr_str) == NULL) {
+        puts("Error: unable to parse destination address");
+        return;
+    }
+    /* parse port */
+    port = (uint16_t)atoi(port_str);
```
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: cast not necessary

File: tests/gnrc_udp/udp.c [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    if (port == 0) {
+        puts("Error: unable to parse destination port");
+        return;
+    }
+
+    data_len = (size_t)atoi(data_len_str);
```
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: cast not necessary

File: tests/gnrc_udp/udp.c [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    uint16_t port;
+
+    /* check if server is already running */
+    if (server.target.pid != KERNEL_PID_UNDEF) {
+        printf("Error: server already running on port %" PRIu32 "\n",
+               server.demux_ctx);
+        return;
+    }
+    /* parse port */
+    port = (uint16_t)atoi(port_str);
```
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: cast

File: tests/gnrc_udp/udp.c [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    if (strcmp(argv[1], "send") == 0) {
+        uint32_t num = 1;
+        uint32_t delay = 1000000;
+        if (argc < 5) {
+            printf("usage: %s send <addr> <port> <bytes> [<num> [<delay in us>]]\n",
+                   argv[0]);
+            return 1;
+        }
+        if (argc > 5) {
+            num = (uint32_t)atoi(argv[5]);
```
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: cast

File: tests/gnrc_udp/udp.c [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        }
+        if (argc > 6) {
+            delay = (uint32_t)atoi(argv[6]);
```
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: cast

File: tests/gnrc_udp/udp.c [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+int udp_cmd(int argc, char **argv)
+{
+    if (argc < 2) {
+        printf("usage: %s [send|server|reset]\n", argv[0]);
+        return 1;
+    }
+
+    if (strcmp(argv[1], "send") == 0) {
+        uint32_t num = 1;
+        uint32_t delay = 1000000;
```
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: 1000000LU

File: tests/gnrc_udp/udp.c (hunk 8 above)
Comment (kYc0o) [inline tests/gnrc_udp/udp.c]: This is the same in the gnrc_networking example. It means it should also be removed there?

Comment (kYc0o) [inline tests/gnrc_udp/udp.c]: The same for almost all the comments below.

Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: Does ```atoi()``` ever have to be casted? I grepped the codebase - there are a lot of casts. If they're never needed, it's very easy to remove them using coccinelle.

Comment (miri64) [inline tests/gnrc_udp/udp.c]: `atoi()` returns `int` so I imagine assigning it to a variable of type with smaller range or `unsigned` might throw some warnings.

File: tests/gnrc_udp/main.c (hunk 3 above)
Comment (miri64) [inline tests/gnrc_udp/main.c]: Well the code is from 2015, I just copied it over from my private app repository. But will extend it to 2017 ;-)

File: tests/gnrc_udp/udp.c (hunk 8 above)
Comment (miri64) [inline tests/gnrc_udp/udp.c]: Okay... compiler isn't complaining. Can somebody explain?

**Comment by miri64** [comment]
Addressed comments

File: tests/gnrc_udp/udp.c (hunk 8 above)
Comment (kaspar030) [inline tests/gnrc_udp/udp.c]: See #7050.

**Comment by kYc0o** [comment]
Well now it needs squash.

**Comment by miri64** [comment]
@kaspar030 are you happy with the changes too?

**Comment by smlng** [comment]
@miri64 might be good to have a test.py script with pexpect to automate this (i.e. call `make test`)?

**Comment by miri64** [comment]
@smlng since this tests requires more than one node, I'd like to @kaspar030's framework for that use-case first.

**Comment by miri64** [comment]
Also this doesn't just include release spec 6, but also needed for e.g. [RS8.7](https://github.com/RIOT-OS/Release-Specs/blob/master/08-interop/08-interop.md#description-6)

**Comment by smlng** [comment]
mhm, I thought more in the direction of what you did for `tests/lwip`. That one is fully automated for board native, which is IMO very nice!

**Comment by miri64** [comment]
> mhm, I thought more in the direction of what you did for tests/lwip. That one is fully automated for board native, which is IMO very nice!

Yes, but that is unnecessarily complicated and currently only works on native. @kaspar030's framework is supposed to be far better than that ;-).

**Comment by aabadie** [comment]
Hey, this one is very useful for the next release. @miri64 @kaspar030, What's missing ?

**Comment by miri64** [comment]
> Hey, this one is very useful for the next release. @miri64 @kaspar030, What's missing ?

An ACK by @kaspar030 ;-)

**Comment by aabadie** [comment]
> An ACK by @kaspar030 ;-)

ping @kaspar030 :)

**Comment by miri64** [comment]
@kaspar030 do we want to merge this at today's Hack'n'ACK?

**Comment by kaspar030** [comment]
ACK.

**Pull request merged**